core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java

```diff
@@ -116,6 +117,7 @@
     private final String message;
     private final Throwable failure;
     private final int failedAllocations;
+    private final Optional<Decision> lastAllocationDecision; // the last allocation decision take for this shard
```

Comment:
We are only interested in Decision.Type, so let's only store that one in UnassignedInfo. Putting the whole Decision object into the cluster state (with explanations and all) seems wasteful to me.